Add javascript to parent theme

matt-bernhardt opened the pull request:
This adds the relevant javascript to the parent theme, following a similar process to #34:

* Copy-paste materials from source repository in one commit
* Use subsequent commits to make adjustments for this new application

The adjustments phase of work has a few different goals. Primary among them is compensating for the loss of the Grunt build process that we've been using until now. This is accomplished by making greater use of `wp_register_script`, defining each of the individual files that went into the five compiled bundles.

Please note, additionally, that we are effectlvely now using _six_ bundles, rather than five. A close inspection of the old `concat.js` file revealed that there were a cluster of files that were used on _every_ page, being included in both the `home` and `production` bundles. I've diagrammed the relationship between files and their bundles as part of LM-28 here:

[Grid of js files and bundles](https://mitlibraries.atlassian.net/102e00b9-5d87-41e5-97ce-cac671819c28)

Other changes are to remove some files that are either not used at all, have been retired, or moved to other parts of the application (such as the hours-loader.js, which is now part of the hours plugin).

Some javascript files had hard-coded paths to install locations that are no longer valid. Those have been replaced with a newly-defined javascript constant that is written by `wp_add_inline_script()`.

There are reference URLs that you can use to verify which files are loaded where in the linked Jira ticket:

[LM-61 - Implement system of registration and enqueuing javascript in themes](https://mitlibraries.atlassian.net/browse/LM-61)

## Developer

### Secrets

- [ ] All new secrets have been added to Pantheon tiers
- [ ] Relevant secrets have been updated in Github Actions
- [ ] All new secrets documented in README
- [x] No new secrets are defined

### Documentation

- [ ] Project documentation has been updated
- [x] No documentation changes are needed

### Accessibility

- [ ] ANDI or Wave has been run in accordance to
      [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened as new
      issues (link to those issues in the Pull Request details above)

### Stakeholder approval

- [ ] Stakeholder approval has been confirmed
- [x] Stakeholder approval is not needed

### Dependencies

YES dependencies are updated (not through Composer, but third-party javascript is added to the parent theme)


## Code Reviewer

- [x] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] The changes have been verified
- [ ] The documentation has been updated or is unnecessary
- [ ] New dependencies are appropriate or there were no changes

matt-bernhardt committed:
Copy-paste js and fonts folders, plus function

** Why are these changes being introduced:

* The first step in bringing in the javascript for the parent theme is
  to copy-paste what the current theme uses. Future commits will rebuild
  this setup, as some referenced files in this commit will not exist /
  be built under Pantheon.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This copy-pastes the existing js and fonts directories (I'm including
  the fonts folder as these are activated via a JS call).
* It also copy-pastes any referenceto registering or enqueueing scripts
  within functions.php.

** Document any side effects to this change:

* Including the fonts is a bit of a side effect, but I'm justifying it
  to myself because these fonts need to be brought in, but couldn't be
  loaded with the stylesheets without the JS call for license compliance.

matt-bernhardt committed:
Reorganize functions.php prior to tackling build

** Why are these changes being introduced:

* The way that we are referencing javascript files in the function
  setup_scripts_styles() is a little disorganized, and ideally would be
  a bit more organized.
* Having a separate commit for this, prior to tackling the teasing apart
  of the build step / concat.js step that is coming, may make things
  easier to follow.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This applies a bit of cleaning to the order of script registration,
  so that third-party libraries (which are usually referenced by locally-
  written scripts) appear first, followed by the (for-now) compiled
  local javascript.

(The next commits will tease apart this compiled javascript)

** Document any side effects to this change:

* Care will be needed when teasing apart the compilation, because there
  are some wildcard references to the js/libs/ directory - and this commit
  changes what is in that folder.

matt-bernhardt committed:
Re-implement maps javascript

** Why are these changes being introduced:

* We need to update how we load javascript from the maps bundle, now
  that we do not have Grunt to build it on the server.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* The maps bundle is super simple, with only a single file and then two
  related libraries (infobox and google maps). This updates the
  references to those files to have the proper relationship and loading.

** Document any side effects to this change:

* None

matt-bernhardt committed:
Re-implement search javascript

** Why are these changes being introduced:

* We need to update how we load javascript from the search bundle.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* The search bundle includes three locally-written files (search,
  search-ie, and ga_discovery). This declares search.js to be the main
  file, with the other two as its dependencies.
* The bundle is loaded by name on both the site homepage and also the
  search page, as both pages have the multisearch widget.

** Document any side effects to this change:

* Eventually, this javascript bundle should probably be moved to the
  multisearch widget itself, rather than kept in the parent theme. That
  should probaby be a separate ticket, however - so I'm electing not to
  do so here.

matt-bernhardt committed:
Re-implement hours javascript

** Why are these changes being introduced:

* We need to update how we load javascript from the hours bundle.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* The hours bundle is relatively straightforward, with six files (three
  libraries, and three written locally). This change moves two of the
  libraries into js/libs for consistency, and moves two of the local
  files into js/ from subdiretories.
* One of the libraries, glDatePicker, was already registered elsewhere,
  so we don't need to introduce it in this commit.

** Document any side effects to this change:

* We are removing two license files which I think were distributed
  along with the jquery libraries that are moving in this commit.
  Ultimately this will need to be something that is addressed during
  dependency management, but for now I'm chosing to prioritize putting
  all the js libraries in a common folder and removing anything else.

matt-bernhardt committed:
Re-implement homepage javascript

** Why are these changes being introduced:

* We need to update how we load javascript for the site homepage.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* There are three javascript files that add functionality specifically
  for the site homepage - loading experts, hours, and guides. This
  registers each, with a dependency upon jquery, and enqueues them
  individually for the site homepage (alongside the search JS, which
  was added in a previous commit).

** Document any side effects to this change:

* This removes an old javascript that was meant to load news content on
  the front page. We had disabled this some time ago because of
  performance reasons, but never deleted the file. Ultimately we will
  probably want to re-implement it, and remove lib/news.php, but that
  will need to wait for future work.

matt-bernhardt committed:
Implement interior javascript

** Why are these changes being introduced:

* We need to update how we load javascript for interior pages, and in
  this case introduce a new bundle.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* A close reading of concat.js from the source theme repo indicats that
  there are two javascript files which are only ever loaded on interior
  pages (not the site homepage, but all other pages).
* This commit creates that "interior" bundle, and enqueues it on the
  appropriate pages.

** Document any side effects to this change:

* As mentioned above, this bundle was not defined explicitly in our
  Grunt setup, but it emerges from paying close attention to what we
  were doing in the build. HOWEVER - this may be a bug, and the two
  files in this bundle may have been intended to be loaded on the site
  homepage in addition to interior pages. Should that prove to be the
  case, we can disband this bundle and move the files to the block of
  materials that are loaded on every page.

matt-bernhardt committed:
Re-implement production javascript

** Why are these changes being introduced:

* We need to update how we load javascript for the production bundle

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This adds the set of files that was formerly compiled into the
  productionJS bundle, declares them as dependencies for the alerts.js
  file, and causes the set to be enqueued on every page.
* The libchat.js file is moved out of the js/libs/ folder, as it is not
  an external library.

** Document any side effects to this change:

* It is a bit awkward to have so many files declared as a dependency for
  alerts.js, as it is not technically true. It might be better to have
  all the files enqueued themselves, which is the model I went with for
  the homepage js (three separate files).

matt-bernhardt committed:
Remove unused javascript files

** Why are these changes being introduced:

* Some javascript files in the js/ foldera appear to neither be
  registered nor enqueued by the theme - so I'm attempting to remove
  them. We should watch the console and user testing feedback to watch
  out for things mysteriously breaking or not loading.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This removes those files from js/ which I cannot detect ever being
  loaded by the theme.

** Document any side effects to this change:

* None

matt-bernhardt committed:
Add customizer preview js

** Why are these changes being introduced:

* There is a bit of javascript needed to help the customizer function,
  separate from our bundled js files.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This copy-pastes the customizer function from the parent theme, along
  with minimal changes to allow it to be called successfully.

** Document any side effects to this change:

* None

matt-bernhardt committed:
Add inline script value

** Why are these changes being introduced:

* Certain paths were hard-coded in our javascript, which have changed
  under Pantheon. While we could update the paths to be accurate now, a
  better solution would be to write the appropriate value using a WP
  function, should these paths ever change in the future.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This adds a call to wp_add_inline_script() to functions.php, to set a
  constant for THEME_ROOT. This is added to the page head element so
  that the rest of the theme javascript will have it available (which is
  mostly loaded in the footer).
* Any references to a "wp-content" path throughout our javascript is
  updated to make use of this constant.

** Document any side effects to this change:

* While debugging this, a few other tweaks are made to get things
  working. For example, a dependency on jQuery is added for the hours
  loader, and some code standard flags are resolved.
* Additionally, a sample div is added to the home page to make sure that
  a CSS-based path is being resolved correctly. This markup will be
  overwritten in a future ticket when the real templates are put in
  place.

matt-bernhardt committed:
Remove hours-loader from parent theme

** Why are these changes being introduced:

* The hours plugin already holds the hours-loader.js script, which is
  enqueued as part of the widgets which display hours information. As
  such, we do not need to maintain a separate copy of the script in the
  parent theme.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This removes the hours-loader.js from the parent theme, as well as the
  libraries upon which it depends.

** Document any side effects to this change:

* The one area of uncertainty right now is the hours page, which may
  not yet be supported by the hours plugin (which has focused on widgets
  so far). However, the next ticket will be to put those templates in
  place, so we will know more for sure soon.
* Additionally, even if we do need to pick up support for the hours
  page, it wouldn't make sense to maintain two copies of the hours
  loader javascript.

matt-bernhardt commented:
Looking at the result of CodeClimate, it seems there may be some duplication that can be removed from the javascript. Specifically, there is `js/map.js` and `js/make.googlemap.js`. We load `map.js`, but not `make.googlemap.js` - so I'm going to remove that in a commit now.

matt-bernhardt committed:
Remove make.googlemap.js

JPrevost commented on web/app/themes/mitlib-parent/functions.php:
```diff
@@ -69,9 +72,77 @@ function setup_scripts_styles() {
+	/* Page-specific JS & CSS */
+
+	// Everything other than the site homepage.
+	if ( ! is_front_page() || is_child_theme() ) {
```
I'm curious if we need the `||` condition here or if just `! is_front_page()` is enough.

Digging into this slightly it seems like each theme might implement these methods and it wasn't clear if they were always the same or if they are different for each theme.

If this is the logic we've been using and it works, I'm not specifically asking for a change more than just pondering whether there is a simpler way to write this to make me think less in the future :)

mitlib-wp-network-deploy replied:
IIRC, this block gets inherited by descendant themes, and thus gets called throughout the network. Having only `! is_front_page()` would result in those assets being included on homepages for News, Exhibits, and any other site on the network - not just the overall network homepage.

I'm open to restructuring this, but am afraid that it would nudge us toward conversations that start to look like a site redesign.

JPrevost replied:
Gotcha. Thanks for clarifying and yes let's just leave it and not go down this particular rabbit hole :)

JPrevost commented on web/app/themes/mitlib-parent/js/alerts.js:
```diff
@@ -0,0 +1,127 @@
+		i;
+
+	// For each post...
+	for (i = 0; i < posts.length; i++) {
```
Is this logic looping over every post on every page load to determine if we have any alerts or is it only fed a smaller set of posts to loop over and the condition checking for the presence of alerts is a safety check?

mitlib-wp-network-deploy replied:
It does loop over all posts on every page load - but the number of posts is pretty small (right now there is only one visible at https://libraries.mit.edu/wp-json/wp/v2/posts/ ).

The other conditions are because not every Post record is necessarily an Alert, and also because there is a safety field that must be checked in order to it to really be displayed here. I would like, ultimately, to make Alerts into their own post type - which would IMO allow us to avoid all of this - but we decided a few weeks ago to avoid this work for now.

JPrevost commented on web/app/themes/mitlib-parent/js/sticky/LICENSE.md:
```diff
@@ -1,21 +0,0 @@
-Copyright 2013 Anthony Garand
```
If these license files are alongside the javascript files they refer to I don't think we should remove them.

I understand the desire to move all the libs to a single directory, but that is outweighed by the license explicitly saying we need to include the license file in at least one of the two that are being removed here. Could we create subdirectories for each under `js/libs` so we can retain the license files for these?

matt-bernhardt replied:
Yeah, that's a fair point. I can restructure this to make separate folders inside `js/libs/` - that makes sense to now, IMO

JPrevost commented on web/app/themes/mitlib-parent/functions.php:
```diff
@@ -101,8 +101,12 @@ function setup_scripts_styles() {
 
 	// Register javascript that we've written.
-
+	wp_register_script( 'dev', get_template_directory_uri() . '/js/dev.js', array(), $theme_version, true );
```
We should figure out why this was ever needed. It seems odd to be protecting allowing console.log type events in production and we'd only want to enable that type of logging in dev/local situations in general. As this is currently in place making any changes is out of scope, but it def feels weird.

matt-bernhardt replied:
Looking at the history for this file in the theme repo, this was added in a single commit and then never amended:

https://github.com/MITLibraries/MITlibraries-parent/commit/edabfc2abee99e1642d6a92e7ed417d3c3c2917c

TBH, I'm not sure how many browsers at this point have no console, and I agree with what you're saying about not spewing messages in production.

JPrevost commented on web/app/themes/mitlib-parent/functions.php:
```diff
@@ -101,8 +101,12 @@ function setup_scripts_styles() {
 	// Register javascript that we've written.
-
+	wp_register_script( 'dev', get_template_directory_uri() . '/js/dev.js', array(), $theme_version, true );
+	wp_register_script( 'fonts', get_template_directory_uri() . '/js/fonts.js', array(), $theme_version, true );
```
Not directly related to this line... by why do we have so many font files? (Not asking for a change, just a bit 🙈 )

matt-bernhardt replied:
I don't remember right now why we have so many; it looks like the PR that added these was https://github.com/MITLibraries/MITlibraries-parent/pull/237 - neither Frances nor I discuss there why so many fonts are being added. The NTI Jira project has been archived, but those specific issues referenced don't include mention either.

JPrevost reviewed (commented):
I don't think I have any blocking comments, but I added a few that will help me better understand a couple things that we may want to open tickets on in a backlog.

We're likely good to go without changes but before merging I'm curious on your answers to a few questions. Thanks!

matt-bernhardt commented:
Thanks for these questions. I think the one element that does require more work, based on this, is to split the contents of `/js/libs/` into sub-folders, to preserve license contents. I'll work on this for an additional commit.

matt-bernhardt committed:
Create separate folders for some js/libs

** Why are these changes being introduced:

* This branch of work included grouping some third-party libraries into
  js/libs. Originally this included removing license files because it
  seemed confusing to have everything in a single folder (and we want to
  deal with dependencies better soon anyway). However, this removing of
  license files raised flags during code review.

** Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/lm-61

** How does this address that need:

* This puts the affected libraries into their own subdirectories under
  js/libs, which allows us to restore the libraries' license files with
  clear relations among the materials.
* The references to these libraries have been updated in functions.php
* Please note that not all js/libs are split in this way - I don't see
  license files for some of these files in our source repository, so I'm
  only splitting libraries which have license files.

** Document any side effects to this change:

* None
